src/Game/Scripting/ParameterList.cs

```diff
@@ -0,0 +1,189 @@
+                if(Aliases.Read<T>(localAlias, content, ref value))
+                    return value;
+            }
+
+            // 3 - Try to resolve it as a global alias (like every uint as 'backpack' maps to the backpack serial)
+            if (Aliases.Read<T>(_node.Lexeme, ref value))
+                return value;
+
+            // 4 - If neither a scope variable or an alias, convert type based on Lexeme
+            if(typeof(T) == typeof(string))
+                return (T)Convert.ChangeType(_node.Lexeme, typeof(string));
+            else return TypeConverter.To<T>(_node.Lexeme);
+        }
+
+        #region Comparison overrided methods
+        public override bool Equals(object obj)
+        {
+            if (obj == null)
+                return false;
+
+            Argument arg = obj as Argument;
+
+            if (arg == null)
+                return false;
+
+            return Equals(arg);
+        }
+
+        public bool Equals(Argument other)
+        {
+            if (other == null)
+                return false;
+
+            return (other._node.Lexeme == _node.Lexeme);
+        }
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
+        }
+        #endregion
+    }
+
+    // A virtualization of the Argument so that arguments may be created and injected in commands on the fly
+    public class VirtualArgument : Argument
+    {
+        // Content of this virtual argument
+        protected string _value;
+
+        public VirtualArgument(string value) : base(null, null)
+        {
+            _value = value;
+        }
+
+        // Generic method to interpreter an argument as the desired type T
+        public override T As<T>(string localAlias = "")
+        {
+            // 1 - Try to resolve it as a local alias (like a ushort set to 0, when a "color", should be ushort.MaxValue)
+            T value = default(T);
+            if (localAlias != string.Empty)
+            {
+                string content = (string)Convert.ChangeType(_value, typeof(string));
+                if (Aliases.Read<T>(localAlias, content, ref value))
+                    return value;
+            }
+
+            // 2 - Try to resolve it as a global alias (like every uint as 'backpack' maps to the backpack serial)
+            if (Aliases.Read<T>(_value, ref value))
+                return value;
+
+            // 3 - If neither a scope variable or an alias, convert type based on stored value
+            if (typeof(T) == typeof(string))
+                return (T)Convert.ChangeType(_value, typeof(string));
+            else return TypeConverter.To<T>(_value);
+        }
+    }
+
+    // Encapsulate AST arguments so that command specific definitions can be evaluated and more easly processed
+    // Using term "parameter" to make difference clear and align term used by scripting languages
+    public class ParameterList
```

Comment:
So I spent a bunch of time trying to use this from the various command handlers to see if it makes the code simpler and I'm only seeing that this actually makes handling the commands more complex. In particular, to do the strict checking of arguments necessary to provide good error messages, calling NextAs isn't particular useful. You need to jump around by index a lot. This can still allow access by index, but if that's the only way it's getting used it may as well stay a basic array.

If you have some good example commands where using this ParameterList makes the implementation simpler I'm happy to look at it again.

Reply:
The ParameterList is a helper class created to:
- removes the need of checking each parameter (optional vs mandatory);
- automates the retrieval of defaults (such as not existing Color mapping to "any" or ushort.MaxValue);
- assist in handling error and propagating an exception with valuable information;
- assist in reading an array parameter;

The simplicity added by ParameterList is proportional to the number of arguments in the command. But it also gets exponentially simpler as common logic in the command can be better used with helped functions.

For example, let's take the UseType command. It can go from this:
Those two helper functionalities for the
```
  private static bool UseType(string command, Argument[] args, bool quiet, bool force)
  {
      if (args.Length < 1)
      {
          throw new RunTimeError(null, "Usage: usetype (graphic) [color] [source] [range or search level]");
      }

      var searchGround = false;

      ushort hue = 0xFFFF;
      Item source = World.Player.FindItemByLayer(Layer.Backpack);

      var graphic = args[0].AsUShort();
      var graphicString = args[0].AsString();

      if (args.Length > 1 && string.Compare(args[1].AsString(), "any", true) != 0)
      {
          hue = args[1].AsUShort();
      }

      if (args.Length > 2)
      {
          if (string.Compare(args[2].AsString(), "ground", true) == 0)
          {
              searchGround = true;
          }
          else
          {
              searchGround = false;
              source = World.Items.Get(args[2].AsSerial());
          }
      }

      var range = args.Length > 3 ? args[3].AsInt() : 0;

      Item item;

      if (searchGround)
      {
          item = World.Player.FindItemByTypeOnGroundWithHueInRange(graphic, hue, range);
      }
      else
      {
          item = World.Player
              .FindItemByLayer(Layer.Backpack)
              .FindItem(graphic, hue);
      }

      if (item != null)
      {
          GameActions.DoubleClick(item);
      }
      else if (!quiet)
      {
          GameActions.Print($"Script Error: Couldn't find '{graphicString}'", type: MessageType.System);
      }

      return true;
  }
```
To this:
```
  private static bool UseType(CommandExecution execution)
  {
      Item item = CmdFindItemByGraphic(
          execution.Params.NextAs<ushort>(ParameterList.Expectation.Mandatory),
          execution.Params.NextAs<ushort>(),
          execution.Params.NextAs<string>(),
          execution.Params.NextAs<int>());
      if(item != null)
          GameActions.DoubleClick(item);
      return (item != null);
  }
```